### DISCUSSION TITLE
Simulation Pipeline Questions

### DISCUSSION THREAD

**Discussion by priesbr1** [python-hydro/pyro2#100] [Q&A] [answered]
I had two quick questions regarding how pyro runs the simulations:

1. To confirm my understanding, the main time loop is `pyro.py/evolve`, which calls `pyro.py/single_step`, which updates the simulation by a timestep dependent on the solver (`sim.compute_timestep` and `sim.evolve`)?
2. In the docs, it says `numba` is a required package, but where exactly is it implemented? I don't see it anywhere in the main time loop I described above, which is where I figured it would be.

**Comment by zingale** [comment, marked as answer]
Yes, the main loop is `run_sim()` in pyro.py which basically calls `single_step()`.

Numba is only used in the compute-intensive kernels.  Look at the output of:

```
git grep numba
```

And you'll see where it is used.  Not every solver needs numba, but for some of the code, if it was too complex to do in slice notation, we explicitly use loops and for those loops, we use the numba JIT compiler.
